### ISSUE TITLE
Select_variable_genes error "ValueError: exog must be a vector"

### ISSUE THREAD

**Issue by Appren893** [pinellolab/STREAM#50]
Hi,

I've been playing around witth STREAM (Bioconda version, through Jupyter) with my dataset for a few days now, and the first time I tried it, it worked without any issues. Since then, I scaled up (at first I tried it with only around 5000 cells, and now I'm running into errors when selecting variable genes.

First, I got an error message saying that "IndexError: tuple index out of range". After some troubleshooting, I found that the "np.std"-function was the cause, as my adata matrix (i.e. adata.X) was sparse. For some reason, np.mean was compatible with a sparse matrix, but np.std wasn't, although that's a different problem.

In order to bypass this, I changed my matrix into a dense one (i.e. "adata.X=adata.X.todense()"), however now I am running into "ValueError: exog must be a vector". Seemingly, this is caused by mean_genes and std_genes being 2-dimensional (i.e. they both have the dimensions 1 x n(genes)), but I do not really know how to fix this, without "manually" running the lowess function, i.e. running the code outside of the function after flattening both arrays. In doing so, the returned "loess_fitted" itself is 2-dimensional and I cannot calculate the "residuals" as the std_genes array is flattened. 

I find it strange that this wouldn't occur everytime that the function is called. I'm using version 0.4.0

Any help is appreciated.
Thanks!

Full traceback:
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-24-6e1c5dd36cdb> in <module>
----> 1 st.select_variable_genes(adata,loess_frac=0.01, percentile=95)

~/anaconda3/envs/myenv/lib/python3.7/site-packages/stream/core.py in select_variable_genes(adata, loess_frac, percentile, n_genes, n_jobs, save_fig, fig_name, fig_path, fig_size)
    514     mean_genes = np.mean(adata.X,axis=0)
    515     std_genes = np.std(adata.X,ddof=1,axis=0)
--> 516     loess_fitted = lowess(std_genes,mean_genes,return_sorted=False,frac=loess_frac)
    517     residuals = std_genes - loess_fitted
    518     XP = np.column_stack((np.sort(mean_genes),loess_fitted[np.argsort(mean_genes)]))

~/anaconda3/envs/myenv/lib/python3.7/site-packages/statsmodels/nonparametric/smoothers_lowess.py in lowess(endog, exog, frac, it, delta, is_sorted, missing, return_sorted)
    131     # same length.
    132     if exog.ndim != 1:
--> 133         raise ValueError('exog must be a vector')
    134     if endog.ndim != 1:
    135         raise ValueError('endog must be a vector')

ValueError: exog must be a vector

**Comment by huidongchen** [comment]
Hi, thanks for the feedbacks. Unfortunately this is also something new to me. Would you mind sharing a minimal example that can reproduce the errors?(You can simply rename your cell and gene names if the data is not public yet) I m happy to look into that

**Comment by Appren893** [comment]
Hi, 

Thanks for your quick reply! I actually found the issue and solved it. When you call adata.X.todense(), it converts adata.X to a dense matrix, not a dense array. As stated in the documentation (https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.matrix.html), "A matrix is a specialized 2-D array that retains its 2-D nature through operations.", meaning that when np.std or np.mean is called, they also return a matrix. Consequently, the LOWESS function returns an error as exog.ndim = 2. I fixed it by calling adata.X=np.array(adata.X.todense()), which resulted in the mean and std being 1D and select_variable_genes thus behaved as expected. I guess I did not have that issue when I tried it the first time because the matrix was already dense due to the lower cell numbers. Perhaps that could be added to the code to prevent it happening in the future?

Sincerely,
